include/zephyr/bluetooth/mesh/access.h

```diff
@@ -302,6 +300,7 @@ struct bt_mesh_model_op {
 	.groups = (uint16_t []) BT_MESH_MODEL_GROUPS_UNASSIGNED(_grps),				\
 	.groups_cnt = _grps,									\
 	.user_data = _user_data,								\
+	.ctx = &(struct bt_mesh_model_ctx){ 0 },						\
```

Comment:
is this compatible with C++?

Reply:
I have not verified C++, but I believe this usage already exists before.

Reply:
I'm sorry, I have never compiled Bluetooth in C++ mode.

Reply:
How？

Reply:
This is indeed not compatible with C++.

Reply:
We have already this issue as `keys` and `groups` are also initialized as compound literals. After discussing with @alwa-nordic, we propose surrounding this problematic defines with `#ifndef __cplusplus`. C++ users can have a C-file (compiled with a C-compiler) with these macros to define variables, and access them in C++ code using `extern`.

Reply:
So, what do I need to do in this PR?

Reply:
@gmarull are you fine with the proposal?

Reply:
@gmarull Please take a look.

Reply:
It's ok to hide stuff from C++ if not compatible (also update API docs). I'd say this is not a great API design choice, though.

Reply:
I looked at this again. Internet says that GCC and clang supports compound literals in C++ as an extension (e.g.: https://gcc.gnu.org/onlinedocs/gcc/Compound-Literals.html). To make GCC complain about compound literal use in C++ code I had to enable `-pedantic` option for GCC, but if I do this in a zephyr app (tested with mesh_provisioner converting main.c to main.cpp), I get C99-related issues in zephyr code. The construction is unique for C and if it is used with C++, the compiler will either use the extension or will warn about the usage of compound literals. We used compound literals since the beginning (6-years old commit: https://github.com/zephyrproject-rtos/zephyr/blob/52cc3b7d00c0e6cf18d0f591fbe771af96802b04/include/bluetooth/mesh/access.h#L162) and this was not a problem before (I remember some customer using C++ but having no issues with this). So I am uncertain that we should anything. At most I can add a disclaimer to the API description, but don't think I need to wrap these macros with ifdefs. @gmarull

Reply:
What about this:
```
diff --git a/include/zephyr/bluetooth/mesh/access.h b/include/zephyr/bluetooth/mesh/access.h
index 07b86ecd9b..ca7655fb73 100644
--- a/include/zephyr/bluetooth/mesh/access.h
+++ b/include/zephyr/bluetooth/mesh/access.h
@@ -270,7 +270,7 @@ struct bt_mesh_model_op {
        .groups_cnt = _grps,                                                    \
        .op = _op,                                                              \
        .cb = _cb,                                                              \
-       .ctx = &(struct bt_mesh_model_ctx){ 0 },                                \
+       .ctx = (struct bt_mesh_model_ctx[]){ { 0 }, },                          \
        .user_data = _user_data,                                                \
 }
```

Reply:
Apart from the compound literal support being not part of the ISO C++, the GNU extension allowing its usage in C++ specifies "In C++, a compound literal designates a temporary object that only lives until the end of its full-expression" (i.e. for C++ with GNU extension, a compound literal is rvalue, whereas it is lvalue in C).

So, even with the GNU extension in C++, the above code is invalid because you cannot take an address of rvalue. In that sense, it should be made clear that the above API cannot be used in C++.

Reply:
Ok, thanks. PR that disables these macros for C++: https://github.com/zephyrproject-rtos/zephyr/pull/58311